text/0000-impl-trait-expressions.md

```diff
@@ -0,0 +1,148 @@
+- Feature Name: impl-trait-expressions
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Rust closures allow the programmer to create values of anonymous types which
+implement the `Fn*` traits. This RFC proposes a generalisation of this feature
+to other traits. The syntax looks like this:
+
+```rust
+fn main() {
+    let world = "world";
+    let says_hello_world = impl fmt::Display {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            write!(f, "hello {}", world)
+        }
+    };
+
+    println!("{}", says_hello_world);
+}
+```
+
+# Motivation
+[motivation]: #motivation
+
+Sometimes we need to create a once-off value which implements some trait,
+though having to explicitly declare a type in these situations can be
+unnecessarily painful and noisy. Closures are a good example of how
+this problem can be ameliorated by adding the ability to declare once-off
+values of anonymous types.
```

Comment:
The motivation feels a bit thin; it would be good to work in some real world use cases of where it would be beneficial; I'm sure you won't have any problems doing that.